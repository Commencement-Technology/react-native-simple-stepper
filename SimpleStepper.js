import React, { Component, PropTypes } from "react";
import { StyleSheet, Text, TouchableOpacity, Image, View } from "react-native";

export default class SimpleStepper extends Component {
  static propTypes = {
    initialValue: PropTypes.number,
    minimumValue: PropTypes.number,
    maximumValue: PropTypes.number,
    stepValue: PropTypes.number,
    backgroundColor: PropTypes.string,
    tintColor: PropTypes.string,
    underlayColor: PropTypes.string,
    padding: PropTypes.number,
    valueChanged: PropTypes.func,
    tintOnIncrementImage: PropTypes.bool,
    tintOnDecrementImage: PropTypes.bool,
    imageHeight: PropTypes.number,
    imageWidth: PropTypes.number,
    activeOpacity: PropTypes.number,
    disabledOpacity: PropTypes.number,
    incrementImage: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    decrementImage: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  };
  static defaultProps = {
    initialValue: 0,
    minimumValue: 0,
    maximumValue: 10,
    stepValue: 1,
    backgroundColor: "transparent",
    tintColor: "blue",
    valueChanged: null,
    decrementImage: require("./assets/decrement.png"),
    incrementImage: require("./assets/increment.png"),
    tintOnIncrementImage: true,
    tintOnDecrementImage: true,
    padding: 4,
    imageHeight: 36,
    imageWidth: 36,
    activeOpacity: 0.4,
    disabledOpacity: 0.5
  };
  constructor(props) {
    super(props);
    this.state = {
      value: props.initialValue,
      decrementOpacity: 1,
      incrementOpacity: 1,
      hasReachedMin: false,
      hasReachedMax: false
    };
    this.decrementAction = this.decrementAction.bind(this);
    this.incrementAction = this.incrementAction.bind(this);
    this.validateInitialValue = this.validateInitialValue.bind(this);
  }
  componentWillMount() {
    this.validateInitialValue(this.props.initialValue);
  }
  componentWillReceiveProps(nextProps) {
    const { initialValue, stepValue } = this.props;
    if (
      nextProps.initialValue !== initialValue ||
      nextProps.stepValue !== stepValue
    ) {
      this.validateInitialValue(nextProps.initialValue);
    }
  }
  decrementAction() {
    var value = this.state.value;
    var stepValue = this.props.stepValue;
    value -= stepValue;
    this.validateInitialValue(value);
  }
  incrementAction() {
    var value = this.state.value;
    var stepValue = this.props.stepValue;
    value += stepValue;
    this.validateInitialValue(value);
  }
  validateInitialValue(value) {
    var maximumValue = this.props.maximumValue;
    var minimumValue = this.props.minimumValue;
    if (value >= maximumValue) {
      value = maximumValue; // prevent overflow value
      this.setState({
        value: maximumValue,
        hasReachedMax: true,
        hasReachedMin: false,
        decrementOpacity: 1,
        incrementOpacity: this.props.disabledOpacity
      });
    } else if (value <= minimumValue) {
      value = minimumValue; // prevent overflow value
      this.setState({
        value: minimumValue,
        hasReachedMin: true,
        hasReachedMax: false,
        decrementOpacity: this.props.disabledOpacity,
        incrementOpacity: 1
      });
    } else {
      if (this.state.hasReachedMax || this.state.hasReachedMin) {
        this.setState({
          value: value,
          hasReachedMin: false,
          hasReachedMax: false,
          decrementOpacity: 1,
          incrementOpacity: 1
        });
      } else {
        this.setState({
          value: value
        });
      }
    }
    if (this.props.valueChanged) {
      this.props.valueChanged(value);
    }
  }
  tintStyle(status) {
    if (status) {
      return { tintColor: this.props.tintColor };
    }
    return null;
  }
  imageIncrementSrc(src) {
    if (src.length == 0) {
      return require("./assets/increment.png");
    }
    if (typeof src == "string") {
      return { uri: src };
    }
    return src;
  }
  imageDecrementSrc(src) {
    if (src.length == 0) {
      return require("./assets/decrement.png");
    }
    if (typeof src == "string") {
      return { uri: src };
    }
    return src;
  }
  imageStyle(src, width, height) {
    if (typeof src == "string" && src.length > 0) {
      return { width: width, height: height };
    }
    return null;
  }
  render() {
    const {
      imageWidth,
      imageHeight,
      tintOnIncrementImage,
      tintOnDecrementImage,
      incrementImage,
      decrementImage,
      padding,
      tintColor,
      backgroundColor,
      activeOpacity
    } = this.props;
    var tintIncrementStyle = this.tintStyle(tintOnIncrementImage);
    var tintDecrementStyle = this.tintStyle(tintOnDecrementImage);
    var decrementImageSrc = this.imageDecrementSrc(decrementImage);
    var incrementImageSrc = this.imageIncrementSrc(incrementImage);
    var incrementStyle = this.imageStyle(
      incrementImage,
      imageWidth,
      imageHeight
    );
    var decrementStyle = this.imageStyle(
      decrementImage,
      imageWidth,
      imageHeight
    );
    const {
      decrementOpacity,
      incrementOpacity,
      hasReachedMin,
      hasReachedMax
    } = this.state;
    return (
      <View
        style={[
          styles.container,
          { backgroundColor: backgroundColor, borderColor: tintColor }
        ]}
      >
        <TouchableOpacity
          ref={ref => this.decrementButton = ref}
          activeOpacity={activeOpacity}
          style={[
            styles.leftButton,
            { borderColor: tintColor, padding: padding }
          ]}
          onPress={this.decrementAction}
          disabled={hasReachedMin}
        >
          <Image
            style={[
              decrementStyle,
              tintDecrementStyle,
              { opacity: decrementOpacity }
            ]}
            source={decrementImageSrc}
          />
        </TouchableOpacity>
        <TouchableOpacity
          ref={ref => this.incrementButton = ref}
          activeOpacity={activeOpacity}
          style={[
            styles.rightButton,
            { borderColor: tintColor, padding: padding }
          ]}
          onPress={this.incrementAction}
          disabled={hasReachedMax}
        >
          <Image
            style={[
              incrementStyle,
              tintIncrementStyle,
              { opacity: incrementOpacity }
            ]}
            source={incrementImageSrc}
          />
        </TouchableOpacity>
      </View>
    );
  }
}

var styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    justifyContent: "center",
    borderWidth: 1,
    borderRadius: 3,
    overflow: "hidden",
    alignItems: "center"
  },
  leftButton: {
    alignItems: "center"
  },
  rightButton: {
    alignItems: "center",
    borderWidth: StyleSheet.hairlineWidth,
    borderBottomWidth: 0,
    borderTopWidth: 0,
    borderRightWidth: 0
  }
});
