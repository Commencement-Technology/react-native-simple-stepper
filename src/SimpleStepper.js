import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import { StyleSheet, View, Text } from 'react-native';
import Step from './Step';
import ImageView from './ImageView';
const STEP = {
  increment: 'increment',
  decrement: 'decrement',
};

export default class SimpleStepper extends Component {
  static propTypes = {
    value: PropTypes.number,
    initialValue: PropTypes.number,
    minimumValue: PropTypes.number,
    maximumValue: PropTypes.number,
    stepValue: PropTypes.number,
    valueChanged: PropTypes.func,
    activeOpacity: PropTypes.number,
    disabledOpacity: PropTypes.number,
    incrementImage: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    decrementImage: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    disabled: PropTypes.bool,
    renderDecrement: PropTypes.func,
    renderIncrement: PropTypes.func,
    wraps: PropTypes.bool,
    onMin: PropTypes.func,
    onMax: PropTypes.func,
    onIncrement: PropTypes.func,
    onDecrement: PropTypes.func,
    showText: PropTypes.bool,
    renderText: PropTypes.func,
    textStyle: PropTypes.object,
    containerStyle: PropTypes.object,
    separatorStyle: PropTypes.object,
    incrementStepStyle: PropTypes.object,
    decrementStepStyle: PropTypes.object,
    incrementImageStyle: PropTypes.object,
    decrementImageStyle: PropTypes.object,
  };
  static defaultProps = {
    value: 0,
    initialValue: 0,
    minimumValue: 0,
    maximumValue: 10,
    stepValue: 1,
    valueChanged: () => {},
    decrementImage: require('./assets/decrement.png'),
    incrementImage: require('./assets/increment.png'),
    activeOpacity: 0.4,
    disabledOpacity: 0.5,
    disabled: false,
    renderDecrement: undefined,
    renderIncrement: undefined,
    wraps: false,
    onMin: () => {},
    onMax: () => {},
    onIncrement: () => {},
    onDecrement: () => {},
    showText: true,
    renderText: undefined,
    textStyle: {
      padding: 8,
      fontSize: 20,
      fontWeight: 'bold',
      color: 'blue',
    },
    containerStyle: {
      backgroundColor: 'transparent',
      flexDirection: 'row',
      borderWidth: 2,
      borderRadius: 8,
      overflow: 'hidden',
      alignItems: 'center',
      borderColor: 'blue',
    },
    separatorStyle: {
      width: StyleSheet.hairlineWidth,
      backgroundColor: 'blue',
      height: '100%',
    },
    incrementStepStyle: {
      padding: 8,
    },
    decrementStepStyle: {
      padding: 8,
    },
    incrementImageStyle: {
      height: 36,
      width: 36,
    },
    decrementImageStyle: {
      height: 36,
      width: 36,
    },
  };
  constructor(props) {
    super(props);
    this.state = {
      changed: false,
    };
  }
  componentDidMount() {
    const { initialValue, value } = this.props;
    // compare value and initialValue if different invoke valueChanged.
    this.validateValue(initialValue, this.props, value !== initialValue);
  }
  componentWillReceiveProps(nextProps) {
    const { value, initialValue, stepValue, minimumValue, maximumValue, disabled } = this.props;
    if (nextProps.initialValue !== initialValue) {
      this.validateValue(nextProps.initialValue, nextProps, nextProps.value !== nextProps.initialValue);
    } else if (nextProps.disabled !== disabled || nextProps.stepValue !== stepValue) {
      this.validateValue(nextProps.value, nextProps);
    } else if (nextProps.minimumValue !== minimumValue || nextProps.maximumValue !== maximumValue) {
      const isValidNextMin = nextProps.minimumValue < maximumValue;
      const isValidNextMax = nextProps.maximumValue > minimumValue;
      if (isValidNextMin && isValidNextMax) {
        this.validateValue(nextProps.value, nextProps);
      }
    } else if (nextProps.value !== value) {
      this.validateValue(nextProps.value, nextProps);
    }
  }
  decrementAction = () => {
    const { value, stepValue, onDecrement } = this.props;
    const nextValue = value - stepValue;
    this.validateValue(nextValue, this.props, true, onDecrement);
  };
  incrementAction = () => {
    const { value, stepValue, onIncrement } = this.props;
    const nextValue = value + stepValue;
    this.validateValue(nextValue, this.props, true, onIncrement);
  };
  validateValue = (value, props, changed = false, onAction = () => {}) => {
    const { minimumValue, maximumValue, wraps, valueChanged, onMin, onMax } = props;
    const { hasReachedMin, hasReachedMax } = this._getHasMinMax(value);
    if (value > maximumValue) {
      value = wraps ? minimumValue : maximumValue;
    } else if (value == maximumValue) {
      value = maximumValue;
    } else if (value < minimumValue) {
      value = wraps ? maximumValue : minimumValue;
    } else if (value == minimumValue) {
      value = minimumValue;
    }
    if (hasReachedMin) {
      onMin(value);
    }
    if (hasReachedMax) {
      onMax(value);
    }
    onAction(value);
    if (changed) {
      this.setState({ changed });
      valueChanged(value);
    }
  };
  _getHasMinMax = value => {
    const { minimumValue, maximumValue, stepValue, wraps } = this.props;
    let hasReachedMax = true;
    let hasReachedMin = true;
    switch (true) {
      case wraps:
        hasReachedMin = false;
        hasReachedMax = false;
        break;
      case stepValue >= 0:
        hasReachedMax = value >= maximumValue;
        hasReachedMin = value <= minimumValue;
        break;
      case stepValue < 0:
        // step value is negative
        // swap the max and min conditions.
        hasReachedMax = value <= minimumValue;
        hasReachedMin = value >= maximumValue;
        break;
    }
    return {
      hasReachedMax,
      hasReachedMin,
    };
  };
  _getImageSource = (type, source) => {
    if (typeof source == 'string') {
      if (source.length == 0) {
        if (type == STEP.decrement) {
          return require('./assets/decrement.png');
        } else if (type == STEP.increment) {
          return require('./assets/increment.png');
        }
      } else {
        return { uri: source };
      }
    }
    return source;
  };
  _renderImageView = (type, opacity) => {
    let style = {};
    let imageSource = null;
    switch (type) {
      case STEP.increment:
        const { renderIncrement, incrementImageStyle, incrementImage } = this.props;
        if (renderIncrement) {
          return renderIncrement(this.props);
        }
        style = incrementImageStyle;
        imageSource = incrementImage;
        break;
      case STEP.decrement:
        const { renderDecrement, decrementImageStyle, decrementImage } = this.props;
        if (renderDecrement) {
          return renderDecrement(this.props);
        }
        style = decrementImageStyle;
        imageSource = decrementImage;
        break;
    }
    return <ImageView style={style} opacity={opacity} source={this._getImageSource(type, imageSource)} />;
  };
  _renderText = (value, renderText, textStyle) => {
    if (renderText) {
      return renderText(value);
    }
    return <Text style={textStyle}>{value}</Text>;
  };
  render() {
    const {
      value,
      activeOpacity,
      disabled,
      disabledOpacity,
      showText,
      renderText,
      containerStyle,
      separatorStyle,
      textStyle,
      incrementStepStyle,
      decrementStepStyle,
    } = this.props;
    const { hasReachedMin, hasReachedMax } = this._getHasMinMax(value);
    const decrementOpacity = hasReachedMin || disabled ? disabledOpacity : 1;
    const incrementOpacity = hasReachedMax || disabled ? disabledOpacity : 1;
    return (
      <View style={containerStyle}>
        <Step
          style={decrementStepStyle}
          activeOpacity={activeOpacity}
          onPress={this.decrementAction}
          disabled={hasReachedMin || disabled}
          renderImage={() => this._renderImageView(STEP.decrement, decrementOpacity)}
        />
        {showText && <View style={separatorStyle} />}
        {showText && this._renderText(value, renderText, textStyle)}
        <View style={separatorStyle} />
        <Step
          style={incrementStepStyle}
          activeOpacity={activeOpacity}
          onPress={this.incrementAction}
          disabled={hasReachedMax || disabled}
          renderImage={() => this._renderImageView(STEP.increment, incrementOpacity)}
        />
      </View>
    );
  }
}
