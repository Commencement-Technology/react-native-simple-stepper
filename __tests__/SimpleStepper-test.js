import {Image} from 'react-native';
import React from 'react';
import SimpleStepper from '../SimpleStepper';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';

test('renders correctly', () => {
  const tree = renderer.create(<SimpleStepper />).toJSON();
  expect(tree).toMatchSnapshot();
});
test('test incrementAction', () => {
  const wrapper = shallow(<SimpleStepper />);
  wrapper.instance().incrementAction();
  expect(wrapper.state().value).toBe(1);
});
test('test decrementAction', () => {
  const wrapper = shallow(<SimpleStepper />);
  wrapper.instance().decrementAction();
  expect(wrapper.state().value).toBe(0);
  expect(wrapper.state().hasReachedMin).toBe(true);
});
test('test incrementAction with wraps', () => {
  const wrapper = shallow(<SimpleStepper wraps={true} initialValue={10} />);
  wrapper.instance().incrementAction();
  expect(wrapper.state().value).toBe(0);
});
test('test decrementAction with wraps', () => {
  const wrapper = shallow(<SimpleStepper wraps={true} />);
  wrapper.instance().decrementAction();
  expect(wrapper.state().value).toBe(10);
});
test('state values to match expected', () => {
  const wrapper = shallow(<SimpleStepper />);
  expect(wrapper.state().value).toEqual(0);
  expect(wrapper.state().decrementOpacity).toBe(0.5);
  expect(wrapper.state().incrementOpacity).toBe(1);
  expect(wrapper.state().hasReachedMin).toBe(true);
  expect(wrapper.state().hasReachedMax).toBe(false);
  expect(wrapper.state().stepValue).toEqual(1);
});
test('validate max value to be true', () => {
  const wrapper = shallow(<SimpleStepper />);
  wrapper.instance().validateValue(10, 0, 10, false, 1);
  expect(wrapper.state().value).toBe(10);
  expect(wrapper.state().hasReachedMax).toBe(true);
});
test('tintColor to be blue', () => {
  const wrapper = shallow(<SimpleStepper />);
  const tintStyle = wrapper.instance().tintStyle(true);
  expect(tintStyle).toEqual({tintColor: 'blue'});
});
test('tintColor to be null', () => {
  const wrapper = shallow(<SimpleStepper />);
  const tintStyle = wrapper.instance().tintStyle(false);
  expect(tintStyle).toBeNull();
});
test('increment type imageSrc to be uri object', () => {
  const wrapper = shallow(<SimpleStepper />);
  const imageSrc = wrapper
    .instance()
    .imageSrc('https://facebook.github.io/react/img/logo_og.png', 'increment');
  expect(imageSrc).toEqual({
    uri: 'https://facebook.github.io/react/img/logo_og.png',
  });
});
test('increment type imageSrc to be local asset number', () => {
  const wrapper = shallow(<SimpleStepper />);
  const imageSrc = wrapper.instance().imageSrc('', 'increment');
  expect(imageSrc).toEqual(1);
});
test('increment type imageSrc to be null', () => {
  const wrapper = shallow(<SimpleStepper />);
  const imageSrc = wrapper.instance().imageSrc(null, 'increment');
  expect(imageSrc).toBeNull();
});
test('decrement type imageSrc to be local asset number', () => {
  const wrapper = shallow(<SimpleStepper />);
  const imageSrc = wrapper.instance().imageSrc('', 'decrement');
  expect(imageSrc).toEqual(1);
});
test('decrement type imageSrc to be undefined', () => {
  const wrapper = shallow(<SimpleStepper />);
  const imageSrc = wrapper.instance().imageSrc(undefined, 'decrement');
  expect(imageSrc).toBeUndefined();
});
test('unknown type imageSrc to be empty string', () => {
  const wrapper = shallow(<SimpleStepper />);
  const imageSrc = wrapper.instance().imageSrc('', 'unknown');
  expect(imageSrc).toEqual('');
});
test('imageStyle to be null', () => {
  const wrapper = shallow(<SimpleStepper />);
  const imageStyle = wrapper.instance().imageStyle(null, 36, 36);
  expect(imageStyle).toBeNull();
});
test('imageStyle to be style object', () => {
  const wrapper = shallow(<SimpleStepper />);
  const imageStyle = wrapper
    .instance()
    .imageStyle('https://facebook.github.io/react/img/logo_og.png', 69, 69);
  expect(imageStyle).toEqual({width: 69, height: 69});
});
test('renderImage to be Image', () => {
  const wrapper = shallow(<SimpleStepper />);
  const renderImage = wrapper.instance().renderImage('eeeee', {}, {}, 1, 1);
  expect(renderImage).toEqual(
    <Image source={1} style={[{}, {}, {opacity: 1}]} />,
  );
});
test('renderImage to be Mock Function', () => {
  const wrapper = shallow(<SimpleStepper />);
  const renderImage = wrapper.instance().renderImage(jest.fn, {}, {}, 1, 1);
  expect(jest.isMockFunction(renderImage)).toEqual(true);
});
