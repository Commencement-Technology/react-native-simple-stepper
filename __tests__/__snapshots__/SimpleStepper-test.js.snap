// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 3,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "transparent",
        "borderColor": "blue",
      },
    ]
  }
>
  <TouchableOpacity
    activeOpacity={0.4}
    disabled={true}
    onPress={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "borderColor": "blue",
          "padding": 4,
        },
      ]
    }
  >
    <Component>
      <Image
        source={
          Object {
            "testUri": "../../../assets/decrement.png",
          }
        }
        style={
          Array [
            null,
            Object {
              "tintColor": "blue",
            },
            Object {
              "opacity": 0.5,
            },
          ]
        }
      />
    </Component>
  </TouchableOpacity>
  <TouchableOpacity
    activeOpacity={0.4}
    disabled={false}
    onPress={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomWidth": 0,
          "borderRightWidth": 0,
          "borderTopWidth": 0,
          "borderWidth": 0.5,
        },
        Object {
          "borderColor": "blue",
          "padding": 4,
        },
      ]
    }
  >
    <Component>
      <Image
        source={
          Object {
            "testUri": "../../../assets/increment.png",
          }
        }
        style={
          Array [
            null,
            Object {
              "tintColor": "blue",
            },
            Object {
              "opacity": 1,
            },
          ]
        }
      />
    </Component>
  </TouchableOpacity>
</Component>
`;
